ssl/quic/quic_port.c

```diff
@@ -647,9 +659,132 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
+ *
+ * Dynamically allocates |buffer| and stores the size of |buffer| in BUFFER_LEN.
+ * Note that it will also allocate an extra QUIC_RETRY_INTEGRITY_TAG_LEN bytes.
+ * The caller is responsible for freeing |buffer|.
+ *
+ * @param token      Validation token.
+ * @param buffer     Callee will allocate a buffer and store the address here.
+ * @param buffer_len Size of |buffer|.
+ */
+static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
+                                    unsigned char **buffer, size_t *buffer_len)
+{
+    WPACKET wpkt = {0};
+    BUF_MEM *buf_mem = BUF_MEM_new();
+
+    if (buf_mem == NULL || (token->is_retry != 0 && token->is_retry != 1))
+        return 0;
+
+    if (!WPACKET_init(&wpkt, buf_mem)
+        || !WPACKET_memset(&wpkt, token->is_retry, 1)
+        || !WPACKET_memcpy(&wpkt, &token->timestamp,
+                           sizeof(token->timestamp))
+        || (token->is_retry
+            && (!WPACKET_sub_memcpy_u8(&wpkt, &token->odcid.id,
+                                       token->odcid.id_len)
+                || !WPACKET_sub_memcpy_u8(&wpkt, &token->rscid.id,
+                                          token->rscid.id_len)))
+        || !WPACKET_sub_memcpy_u8(&wpkt, token->remote_addr, token->remote_addr_len)
+        || !WPACKET_get_total_written(&wpkt, buffer_len)
+        || !WPACKET_finish(&wpkt)
+        || !BUF_MEM_grow(buf_mem, *buffer_len + QUIC_RETRY_INTEGRITY_TAG_LEN)) {
+        WPACKET_cleanup(&wpkt);
+        BUF_MEM_free(buf_mem);
+        return 0;
+    }
+
+    *buffer = (unsigned char *)buf_mem->data;
+    buf_mem->data = NULL;
+    BUF_MEM_free(buf_mem);
+    *buffer_len += QUIC_RETRY_INTEGRITY_TAG_LEN;
+    return 1;
+}
+
+/**
+ * @brief Parses contents of a buffer into a validation token.
+ *
+ * VALIDATION_TOKEN should already be initalized. Does some basic sanity checks.
+ *
+ * @param token   Validation token to fill data in.
+ * @param buf     Buffer of previously marshaled validation token.
+ * @param buf_len Length of |buf|.
+ */
+static int parse_validation_token(QUIC_VALIDATION_TOKEN *token,
+                                  const unsigned char *buf, size_t buf_len)
+{
+    PACKET pkt;
+
+    if (buf == NULL || token == NULL)
+        return 0;
+
+    token->remote_addr = NULL;
+
+    if (!PACKET_buf_init(&pkt, buf, buf_len)
+        || !PACKET_copy_bytes(&pkt, &token->is_retry, sizeof(token->is_retry))
+        || !(token->is_retry == 0 || token->is_retry == 1)
+        || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->timestamp,
+                              sizeof(token->timestamp))
+        || (token->is_retry
+            && (!PACKET_copy_bytes(&pkt, &token->odcid.id_len,
+                sizeof(token->odcid.id_len))
+                || token->odcid.id_len > QUIC_MAX_CONN_ID_LEN
+                || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->odcid.id,
+                                      token->odcid.id_len)
```

Comment:
If we ever change the type of `id_len` this will start failing. `sizeof` an unsigned char is 1. When marshalling the token we use `WPACKET_sub_memcpy_u8` - so by using `u8` the number of length bytes is explicitly 1, regardless of the type of `id_len`. 

Better would be to use sub-packets for this - which then matches the marshalling method of using sub-packets
```suggestion
            && (!PACKET_get_length_prefixed_1(&pkt, &subpkt)
                || (token->odcid.id_len = PACKET_remaining(&subpkt)) > QUIC_MAX_CONN_ID_LEN
                || !PACKET_copy_bytes(&subpkt, (unsigned char *)&token->odcid.id,
                                      token->odcid.id_len)
```